File: apps/api/src/api/i18n/catalogue.py
"""Reading the shared message catalogue - FR-LOC-001, FR-LOC-001c, FR-UX-007.

    FR-LOC-001  Every screen, error message, email, push notification, PDF
                template, help article and validation message exists in both.
                A missing translation is a release blocker, not a fallback to
                English.

--- One catalogue, read by two runtimes ---

The files under packages/i18n/catalogue/ are the source of every user-facing
string in the product, for the API and for the clients alike. This module
reads exactly the same directory packages/i18n/src/catalogue.ts imports, and
does not keep a copy.

That is worth the small awkwardness of reaching outside apps/api for data. The
alternative - the API owning its error messages and the web owning its labels
- means the same sentence exists twice, is translated twice, and drifts. An
error a user reads on screen and the same error rendered into a PDF have to be
the same sentence in the same words, and the cheapest way to guarantee that is
for there to be only one of it.

The directory is resolved at import, not at first use, and a missing or
unreadable catalogue is a startup failure. A product whose every user-facing
string is missing should refuse to start rather than serve raw message keys.

--- No fallback between languages ---

`translate` RAISES on a key it cannot resolve, and there is no path from a
missing Dutch string to the English one. FR-LOC-001 rules that out explicitly,
and the reason is not tidiness: a fallback makes a missing translation
invisible everywhere somebody might notice it, so the Dutch product ships with
English sentences scattered through it and nobody has a list of them.

Raising is only safe because scripts/check_translations.py (FR-LOC-001d) fails
the build first. The check is what makes the runtime posture defensible.
"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from functools import cache
from pathlib import Path
from typing import Any

from api.config import settings
from api.i18n.language import Language, plural_category

#: Namespace files, in load order. Named rather than globbed: a catalogue file
#: that is present on disk and absent from this list would be silently
#: unavailable to the API while the web app served it happily, and a glob would
#: make that failure depend on what happened to be in the directory.
#: scripts/check_translations.py asserts this list matches the directory.
CATALOGUE_FILES = (
    "common.json",
    "auth.json",
    "capture.json",
    "client.json",
    "errors.json",
    "invoice.json",
    "mobile.json",
    "roles.json",
)

GLOSSARY_FILE = "glossary.json"


def _candidate_directories() -> list[Path]:
    """Where the catalogue might be, most specific first.

    Two real locations, and they are not redundant:

      * the packaged copy, `api/i18n/catalogue`, which `uv build` places in
        the wheel via the `force-include` in pyproject.toml. This is what a
        deployed API reads.
      * the checkout, `packages/i18n/catalogue`, five directories up. This is
        what a developer, the test suite and CI read, and it is the one the
        web app reads too - so the two runtimes are demonstrably on the same
        bytes during development.
    """
    here = Path(__file__).resolve()
    candidates = []
    if settings.message_catalogue_dir:
        candidates.append(Path(settings.message_catalogue_dir))
    candidates.append(here.parent / "catalogue")
    candidates.append(here.parents[5] / "packages" / "i18n" / "catalogue")
    return candidates


def catalogue_directory() -> Path:
    for candidate in _candidate_directories():
        if (candidate / GLOSSARY_FILE).is_file():
            return candidate
    raise RuntimeError(
        "no message catalogue found. Looked in: "
        + ", ".join(str(path) for path in _candidate_directories())
        + ". Every user-facing string comes from it (FR-LOC-001), so the API "
        "does not start without one. Set MESSAGE_CATALOGUE_DIR if it lives "
        "somewhere else."
    )


@dataclass(frozen=True, slots=True)
class GlossaryTerm:
    """FR-LOC-001c. A statutory term that keeps its Dutch form in the English
    UI, with the hover or tap definition the requirement asks for.
    """

    term: str
    keep_dutch: bool
    definition: dict[str, Any]


class MissingMessageError(KeyError):
    """A key the catalogue does not have, a counted message with no count, or
    a placeholder with no value. All three would otherwise reach a user as a
    raw key, the wrong plural form, or a literal `{query}` on screen - and
    FR-UX-007 is explicit that developer-facing strings never reach a user.
    """


def _load() -> tuple[dict[str, dict[str, Any]], dict[str, GlossaryTerm]]:
    directory = catalogue_directory()

    messages: dict[str, dict[str, Any]] = {}
    for filename in CATALOGUE_FILES:
        document = json.loads((directory / filename).read_text(encoding="utf-8"))
        for key, record in document["messages"].items():
            if key in messages:
                # Two files claiming one key means one of them silently loses,
                # and which one would depend on load order. The namespace rule
                # (every key in client.json starts with `client.`) makes this
                # unreachable; it is asserted rather than assumed.
                raise RuntimeError(f"duplicate message key across catalogue files: {key}")
            messages[key] = record

    glossary_document = json.loads((directory / GLOSSARY_FILE).read_text(encoding="utf-8"))
    glossary = {
        term_id: GlossaryTerm(
            term=entry["term"],
            keep_dutch=bool(entry.get("keepDutch", False)),
            definition=entry["definition"],
        )
        for term_id, entry in glossary_document["terms"].items()
    }

    return messages, glossary


MESSAGES, GLOSSARY = _load()


_PLACEHOLDER = re.compile(r"\{(\w+)\}")


def has_message(key: str) -> bool:
    """Whether a key exists, for the rare caller that builds one at runtime."""
    return key in MESSAGES


def translate(key: str, language: Language, /, **params: object) -> str:
    """The message for `key`, in `language`, with `{placeholders}` filled in.

    `language` is positional-only alongside `key` so that a message parameter
    happening to be called `language` cannot shadow it - which is exactly the
    kind of collision a `**params` signature invites.
    """
    record = MESSAGES.get(key)
    if record is None:
        raise MissingMessageError(
            f"no message {key!r} in the catalogue. Add it to "
            f"packages/i18n/catalogue/ in BOTH languages (FR-LOC-001)."
        )

    text = _select(record[language.value], key, language, params)
    return _interpolate(text, key, params)


def _select(text: Any, key: str, language: Language, params: dict[str, object]) -> str:
    if isinstance(text, str):
        return text

    count = params.get("count")
    if not isinstance(count, int | float) or isinstance(count, bool):
        raise MissingMessageError(
            f"{key} is a counted message and needs a numeric `count` parameter; got {count!r}."
        )
    category = plural_category(language, count)
    return str(text[category])


def _interpolate(text: str, key: str, params: dict[str, object]) -> str:
    def replace(match: re.Match[str]) -> str:
        name = match[1]
        if name not in params:
            raise MissingMessageError(
                f"{key} expects a {{{name}}} parameter and was given none. The "
                f"catalogue keeps the same placeholders in both languages, so this "
                f"is missing in both."
            )
        return str(params[name])

    return _PLACEHOLDER.sub(replace, text)


@cache
def glossary_definition(term_id: str, language: Language) -> str | None:
    """FR-LOC-001c's hover or tap definition, or None for an unknown term.

    None rather than a raise: a glossary lookup is decoration around a term
    that is already legible on its own, so a missing entry should cost the
    tooltip and nothing else.
    """
    entry = GLOSSARY.get(term_id)
    if entry is None:
        return None
    text = entry.definition[language.value]
    return text if isinstance(text, str) else str(text["other"])

File: packages/i18n/src/catalogue.ts
/**
 * Reading the message catalogue — FR-LOC-001, FR-LOC-001c, FR-UX-007.
 *
 * The catalogue itself is ../catalogue/*.json and its contract is documented
 * there. This module is the runtime side of it: types, the merged lookup, and
 * interpolation.
 *
 * --- There is no fallback between languages, on purpose ---
 *
 * Nearly every i18n library resolves a missing Dutch string by showing the
 * English one. FR-LOC-001 forbids exactly that — "A missing translation is a
 * release blocker, not a fallback to English" — and the reason is not
 * tidiness. A fallback makes a missing translation invisible in every
 * environment where somebody might notice it, so the Dutch product ships with
 * English sentences scattered through it and nobody has a list of them.
 *
 * So `translate` THROWS on a key it cannot resolve, and the safety net is
 * scripts/check_translations.py (FR-LOC-001d) failing the build first. The
 * order matters: the check is what makes throwing at runtime a safe design
 * rather than a reckless one.
 */

import type { Language } from "./language";
import { pluralCategory } from "./language";

import authFile from "../catalogue/auth.json";
import captureFile from "../catalogue/capture.json";
import clientFile from "../catalogue/client.json";
import commonFile from "../catalogue/common.json";
import customersFile from "../catalogue/customers.json";
import errorsFile from "../catalogue/errors.json";
import invoiceFile from "../catalogue/invoice.json";
import glossaryFile from "../catalogue/glossary.json";
import ledgerFile from "../catalogue/ledger.json";
import mobileFile from "../catalogue/mobile.json";
import onboardingFile from "../catalogue/onboarding.json";
import rolesFile from "../catalogue/roles.json";
import settingsFile from "../catalogue/settings.json";

/** A single form, or the two CLDR categories Dutch and English share. */
export type MessageText = string | { readonly one: string; readonly other: string };

export interface MessageRecord {
  /** Context for whoever writes or reviews the translation. Never rendered. */
  readonly note?: string;
  /** Required when the two languages match — see ../catalogue/README.md. */
  readonly identical?: boolean;
  readonly nl: MessageText;
  readonly en: MessageText;
}

export interface GlossaryTerm {
  /** The term as it is written in Dutch, and in English when kept. */
  readonly term: string;
  /** FR-LOC-001c: this term survives translation unchanged. */
  readonly keepDutch: boolean;
  /** The hover or tap definition the requirement asks for, in both languages. */
  readonly definition: MessageRecord;
}

interface CatalogueFile {
  readonly namespace: string;
  readonly messages: Readonly<Record<string, MessageRecord>>;
}

interface GlossaryFile {
  readonly namespace: string;
  readonly terms: Readonly<Record<string, GlossaryTerm>>;
}

// The JSON files are data, and TypeScript infers a wildly specific literal
// type for each. Widening once here, at the boundary, keeps the assertion in
// one place — and the shapes are checked for real by
// scripts/check_translations.py, which reads the same files without believing
// anything TypeScript says about them.
const FILES: readonly CatalogueFile[] = [
  commonFile as CatalogueFile,
  authFile as CatalogueFile,
  captureFile as CatalogueFile,
  clientFile as CatalogueFile,
  errorsFile as CatalogueFile,
  invoiceFile as CatalogueFile,
  mobileFile as CatalogueFile,
  rolesFile as CatalogueFile,
  // The routed web app's four namespaces (docs/founder-review-2026-09-14.md
  // §5.2). Listed here AND in api.i18n.catalogue.CATALOGUE_FILES;
  // scripts/check_translations.py fails if the two lists differ.
  onboardingFile as CatalogueFile,
  settingsFile as CatalogueFile,
  ledgerFile as CatalogueFile,
  customersFile as CatalogueFile,
];

function mergeMessages(files: readonly CatalogueFile[]): Record<string, MessageRecord> {
  const merged: Record<string, MessageRecord> = {};
  for (const file of files) {
    for (const [key, record] of Object.entries(file.messages)) {
      if (key in merged) {
        // Two files claiming one key means one of them silently loses. The
        // namespace rule (every key in client.json starts with `client.`)
        // makes this unreachable, and it is asserted rather than assumed
        // because the failure would be a message that changes depending on
        // file load order.
        throw new Error(`duplicate message key across catalogue files: ${key}`);
      }
      merged[key] = record;
    }
  }
  return merged;
}

export const MESSAGES: Readonly<Record<string, MessageRecord>> = mergeMessages(FILES);

export const GLOSSARY: Readonly<Record<string, GlossaryTerm>> = (glossaryFile as GlossaryFile)
  .terms;

export type MessageKey = string;

export type MessageParams = Readonly<Record<string, string | number>>;

export class MissingMessageError extends Error {}

/** Whether a key exists, for the rare caller that builds one at runtime. */
export function hasMessage(key: MessageKey): boolean {
  return key in MESSAGES;
}

/**
 * The message for `key`, in `language`, with `{placeholders}` filled in.
 *
 * Throws when the key is unknown, when a counted message is given no `count`,
 * or when a placeholder has no value. All three are bugs that would otherwise
 * reach a user as a raw key, the wrong plural form, or a literal `{query}` on
 * screen — and FR-UX-007 is explicit that developer-facing strings never
 * reach a user.
 */
export function translate(key: MessageKey, language: Language, params: MessageParams = {}): string {
  const record = MESSAGES[key];
  if (record === undefined) {
    throw new MissingMessageError(
      `no message ${JSON.stringify(key)} in the catalogue. Add it to ` +
        `packages/i18n/catalogue/ in BOTH languages (FR-LOC-001).`,
    );
  }

  const text = language === "nl" ? record.nl : record.en;
  return interpolate(select(text, key, language, params), key, params);
}

function select(
  text: MessageText,
  key: MessageKey,
  language: Language,
  params: MessageParams,
): string {
  if (typeof text === "string") return text;

  const count = params.count;
  if (typeof count !== "number") {
    throw new MissingMessageError(
      `${key} is a counted message and needs a numeric \`count\` parameter; ` +
        `got ${JSON.stringify(count)}.`,
    );
  }
  return pluralCategory(language, count) === "one" ? text.one : text.other;
}

const PLACEHOLDER = /\{(\w+)\}/g;

function interpolate(text: string, key: MessageKey, params: MessageParams): string {
  return text.replace(PLACEHOLDER, (_match, name: string) => {
    const value = params[name];
    if (value === undefined) {
      throw new MissingMessageError(
        `${key} expects a {${name}} parameter and was given none. The catalogue ` +
          `keeps the same placeholders in both languages, so this is missing in ` +
          `both.`,
      );
    }
    return String(value);
  });
}

/**
 * FR-LOC-001c's hover or tap definition. Returns `undefined` for an unknown
 * term rather than throwing: a glossary lookup is decoration around a term
 * that is already legible on its own, so a missing entry should cost the
 * tooltip and nothing else.
 */
export function glossaryDefinition(termId: string, language: Language): string | undefined {
  const entry = GLOSSARY[termId];
  if (entry === undefined) return undefined;
  const text = language === "nl" ? entry.definition.nl : entry.definition.en;
  return typeof text === "string" ? text : text.other;
}

export function glossaryTerm(termId: string): GlossaryTerm | undefined {
  return GLOSSARY[termId];
}
